src/Types/TypeRegistry.php

```diff
@@ -0,0 +1,169 @@
+        Type::STRING => StringType::class,
+        Type::DATE => DateType::class,
+        Type::DATE_IMMUTABLE => DateImmutableType::class,
+        Type::KEY => KeyType::class,
+        Type::TIMESTAMP => TimestampType::class,
+        Type::BINDATA => BinDataType::class,
+        Type::BINDATAFUNC => BinDataFuncType::class,
+        Type::BINDATABYTEARRAY => BinDataByteArrayType::class,
+        Type::BINDATAUUID => BinDataUUIDType::class,
+        Type::BINDATAUUIDRFC4122 => BinDataUUIDRFC4122Type::class,
+        Type::BINDATAMD5 => BinDataMD5Type::class,
+        Type::BINDATACUSTOM => BinDataCustomType::class,
+        Type::HASH => HashType::class,
+        Type::COLLECTION => CollectionType::class,
+        Type::OBJECTID => ObjectIdType::class,
+        Type::RAW => RawType::class,
+        Type::DECIMAL128 => Decimal128Type::class,
+        Type::UUID => BinaryUuidType::class,
+        Type::VECTOR_FLOAT32 => VectorFloat32Type::class,
+        Type::VECTOR_INT8 => VectorInt8Type::class,
+        Type::VECTOR_PACKED_BIT => VectorPackedBitType::class,
+    ];
+
+    /** @var array<string, Type> Cache of instantiated Type objects */
+    private array $typeObjects = [];
+
+    private static ?TypeRegistry $sharedInstance = null;
+
+    /**
+     * Register a new type in the type map.
+     *
+     * The name of the type can be a PHP class name used for automatic type detection
+     *
+     * @param non-empty-string        $name
+     * @param class-string<Type>|Type $type
+     */
+    public function register(string $name, string|Type $type): void
+    {
+        if ($type instanceof Type) {
+            $this->typesMap[$name]    = $type::class;
+            $this->typeObjects[$name] = $type;
+        } else {
+            $this->typesMap[$name] = $type;
+            unset($this->typeObjects[$name]);
+        }
+    }
+
+    /**
+     * Checks if exists support for a type.
+     */
+    public function has(string $name): bool
+    {
+        return isset($this->typesMap[$name]);
+    }
+
+    /**
+     * Get a Type instance.
+     *
+     * @throws InvalidTypeException
+     */
+    public function get(string $name): Type
+    {
+        if (! isset($this->typesMap[$name])) {
+            throw InvalidTypeException::invalidTypeName($name);
+        }
+
+        return $this->typeObjects[$name] ??= (new ReflectionClass($this->typesMap[$name]))->newInstanceWithoutConstructor();
+    }
+
+    /**
+     * Determine the database representation of a value based on its PHP type.
+     */
+    public function convertToDatabaseValue(mixed $value): mixed
+    {
+        $type = $this->fromVariable($value);
+
+        if ($type === null) {
+            return $value;
+        }
+
+        return $type->convertToDatabaseValue($value);
+    }
+
+    /**
+     * Get a Type instance based on the type of the passed PHP variable.
+     *
+     * @internal
+     */
+    public function fromVariable(mixed $variable): ?Type
```

Comment:
I would rename this to better indicate that we're guessing a type, and that we're expecting a native value and not a database value.

```suggestion
    public function guessTypeFromValue(mixed $variable): ?Type
```

In future, we could consider adding a separate interface that allows types to indicate whether they can handle a certain value.

Reply:
Method renamed with your good suggestion.

> In future, we could consider adding a separate interface that allows types to indicate whether they can handle a certain value.

That's delicate as we need to have the same detection from property type and from values.